### ISSUE TITLE
Injecting certain services causes O(n) compilation complexity, while injecting Container is O(1)

### ISSUE THREAD

**Issue by ostrolucky** [symfony/symfony#37850]
**Symfony version(s) affected**: 4.4.11

**Description**  
Injecting services with complicated service graphs causes large compilation slowdown compared to simply injecting container. 

This comes as a surprise, since injecting services directly comes as universal recommendation, people are instructed to do it in migration guides from Symfony 3.x to 4.x and is even recommended for performance reasons, while I observe opposite.

**How to reproduce**  
1. Download [project.zip](https://github.com/symfony/symfony/files/5081372/project.zip)
1. `composer install`
1. `time bin/console`

Notice the time. In my case this is around **10s** without xdebug and **1m:5s** with xdebug.

Now do
```diff
Index: project/src/AppBundle/Controller/Admin/Settings/FormFieldTypeController.php
===================================================================
@@ -12,9 +12,9 @@
 class FormFieldTypeController extends AbstractController
 {
     public function __construct(
-        PanelHandlerAll $panelHandlerAll,
-        CustomPanelLoader $customPanelLoader
-//        ContainerInterface $container
+        ContainerInterface $container
     ) {
     }
 }

```

Run `time bin/console` again. Notice the time. In my case this is around **3s** without xdebug and around **15s** with xdebug.

Now, why did I mention this is O(n) vs O(1) problem:
In `Resources/config/services/_default.yml` file is `c1`..`c26` service defined. In case of injecting service container, it doesn't really matter how many of them there are, compilation speed is pretty much constant. However in other case, speed is very much dependant on number of these service definitions.

**Additional context**  
I originally raised this issue on Slack. @nicolas-grekas encouraged me to create this issue ;) I was working on converting the retrieval of our public services to injecting them directly, as officially recommended as Symfony best practice. In middle of this I discovered compilation times got much worse than originally, as opposed to just injecting the ContainerInterface as was done before. I have spent lot of time gutting our application so I can provide reproducer, I hope it's useful, even though I know it's not perfect.

Possibly related: #37540. However, marking the services private has no impact on performance in my case.

Blackfire link: https://blackfire.io/profiles/17769187-521c-48d3-8600-3de9adece809/graph (this is from before I gutted the application into reproducer)

**Comment by xgrn** [comment]
Hi. I can confirm the issue. We used the whole container injection on 2.8 then updated all our services to exact service injections following the 3.4 migration guides. As we have a huge amount of services and a complex dependency graph it takes > 1h for our container to be built now. If I turn circular dependency analyzing off it's dropped to a minute.

Making all the services private doesn't help.

**Comment by ostrolucky** [comment]
> If I turn circular dependency analyzing off it's dropped to a minute.

Could you explain this more? How was this done?

**Comment by xgrn** [comment]
Straight and rude. I just commend the following lines in the PhpDumper::dump()

```php

        (new AnalyzeServiceReferencesPass(false, !$this->getProxyDumper() instanceof NullDumper))->process($this->container);
        $checkedNodes = [];
        $this->circularReferences = [];
//        foreach ($this->container->getCompiler()->getServiceReferenceGraph()->getNodes() as $id => $node) {
//            if (!$node->getValue() instanceof Definition) {
//                continue;
//            }
//            if (!isset($checkedNodes[$id])) {
//                $this->analyzeCircularReferences($id, $node->getOutEdges(), $checkedNodes);
//            }
//        }
        $this->container->getCompiler()->getServiceReferenceGraph()->clear();
        $checkedNodes = [];

```

**Comment by timglabisch** [comment]
We've the same issue. these lines are taking 7m in my setup. (v4.4.15 with xdebug2).

**Comment by timglabisch** [comment]
i have taken a closer look at the problem, because the compile time has increased dramatically. in an old branch it was 16 seconds, suddenly it was 7m

it looks like the logic for resolving the CircularReferences is not very efficient, because he always goes through all services and checks the complete path. I don't understand the implementation 100% yet but probably there is some room for improvement. In my case a service was tagged with monolog.processor, which needed doctrine in its dependencies. doctrine in turn uses the logger. This led to entries in the PhpDumber::circularReferences . But there was no circularReference. 

This led to a deep search path which was run with o(n) every time the logger was needed somewhere.
I would recommend everyone to set a breakpoint in the addCircularReferences. A deep $currentPath seems to be an indication for performance problems.

The solution here is to use $container->get() directly, because the complexity here is only o(n). But the call must not take place in the controller.
<img width="1589" alt="Screen Shot 2020-10-29 at 11 44 32" src="https://user-images.githubusercontent.com/689269/97559198-66514100-19dd-11eb-8b29-cc9687c93dac.png">

**Comment by jderusse** [comment]
The title and the description if this issue are **misleading** and an important detail is missing:

Your are talking about time to compile the container, which occurred only **once in a lifetime** of production environment and only when a service change in dev environment.

While, in my opinion, what matters is the performance in a production environments, in that case, we should compare performance:
```diff
-$this->container->get('service');
+$this->service;
```

I don't have number, but I'm pretty sure that "fetching a property", is faster than "fetching a property + calling a method that lookup in a hashmap". And **YES** injecting dependencies is **FASTER** than injecting the Container. 

That's said, there maybe room to optimize that `analyzeCircularReferences` part of the code.

In the case of the project uploaded by @ostrolucky there are a lot of CircularReferences, and the Dumper spent a lot of time checking each one by one:

For instance, here are 2 circular reference:

> AppBundle\Panel\PanelHandlerAll => AppBundle\Panel\Panels\Application\Labels => AppBundle\Panel\Component\Renderer\TwigRenderer => twig => AppBundle\Panel\PanelHandlerAll

and

> AppBundle\Panel\PanelHandlerAll => AppBundle\Panel\Panels\Application\Labels => panel_handler.application => AppBundle\Panel\Panels\Application\Labels

There are **72 occurences** like this, all connected to the Service C1.
But worse, those references are also injected in C1, C2, C3, ....,  C25, the dumper spent about 200 ms in each service in connectCircularReferences and call the method `addCircularReferences` 7 993 126 times !!!

Those circular references are not a dead end for you, because at the end, they are not injected via the constructor, but symfony have to check theme one by one in case of circular happens with constructor injection.

**Comment by jderusse** [comment]
Could you please have a look to this PR to check if perfs are better? https://github.com/symfony/symfony/pull/38882

**Comment by ostrolucky** [comment]
> The title and the description if this issue are misleading and an important detail is missing:

They are not, both explicitly state *compilation*.

> While, in my opinion, what matters is the performance in a production environments, in that case, we should compare performance:

Both matter. 10sec+ compilation times very much kill dev productivity. Especially for code that is not even used in request. There is not many PHP developers who would be willing to bear such times. This is a Symfony specific problem, I don't know about any other PHP project doing these compilations.



And yes we all know this is because of circular references, it's visible both from blackfire link I posted in description and from the code @xgrn posted. Point is, everybody, including documentation states that you should just blindly inject everything, it doesn't mention circular references, then people who spend lot of effort to move everything to constructor injections get burned. This made me actually miss good ol days of symfony container complaining about circular references. At least I would notice problem early and would inject services in those cases via other means, eg. setter.



> Could you please have a look to this PR to check if perfs are better? #38882

Thanks for looking into this! I'm using Symfony 4.4, but during my testing of your patch in my project looks to lead to some sort of infinite loop. I'll keep it running for some more time and see.

edit: yep, still didn't stop after 5 min from usual few 5 sec. so I'm calling it quits

**Comment by jderusse** [comment]
> Thanks for looking into this! I'm using Symfony 4.4, but during my testing of your patch in my project looks to lead to some sort of infinite loop. I'll keep it running for some more time and see.
> 
> edit: yep, still didn't stop after 5 min from usual few 5 sec. so I'm calling it quits

Hmm weird, initial implemented the patch on 4.4 (code here: https://github.com/jderusse/symfony/blob/perf-circular-44/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php) and rebased onto 5.5. Are you using the same project uploaded in the issue?

**Comment by nicolas-grekas** [comment]
> This made me actually miss good ol days of symfony container complaining about circular references. At least I would notice problem early and would inject services in those cases via other means, eg. setter.

I'm sad reading such comments about the supposed "good old days". There is no need to be emotional, but we are human beings. About these good ol' days: nothing changed, circular refs by constructors are still detected using the same logic as before, and they're detected very quickly since always.

The logic we're talking about is meant to detect loops created "via other means, eg. setter.". These loops *are* solvable, but they require precise knowledge about the way the loop is structured, in order to decide about the *order* of the calls to the setters.

I agree the current algorithm takes too long. It's not `O(n)` by the way but more like `O(n²)` or maybe a higher order. We should find a better way. Thanks @jderusse for giving it a try!
